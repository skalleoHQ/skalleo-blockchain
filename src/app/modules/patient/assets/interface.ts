export interface CreatePatientAccountAssetProps {
	patientIdentificationNumber: string;
	areaCode: string;
	username: string;
}