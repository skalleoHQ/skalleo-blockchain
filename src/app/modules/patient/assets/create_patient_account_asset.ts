import { BaseAsset, ApplyAssetContext, ValidateAssetContext } from 'lisk-sdk';

const {
	VALID_PATIENT_DOMAIN,
	createPatientAccount,
	getAllPatientAccounts,
	setAllPatientAccounts,

} = require("../data/utils")



export class CreatePatientAccountAsset extends BaseAsset {
	public name = 'createPatientAccount';
  	public id = 1;

  // Define schema for asset
	public schema = {
    $id: 'patient/createPatientAccount-asset',
		title: 'CreatePatientAccountAsset transaction asset for patient module',
		type: 'object',
		required: ['patientIdentificationNumber', 'areaCode', 'username'],
		properties: {
			patientIdentificationNumber: {
				dataType: 'string',
				fieldNumber: 1,
			},
			username: {
				dataType: 'string',
				fieldNumber: 2,
			},
			areaCode: {
				dataType: 'string',
				fieldNumber: 3,
			},
		},
  	};

	public validate({ asset }: ValidateAssetContext<{}>): void {
		// Verify if client fill patient Identification Number box
		if (!asset.patientIdentificationNumber) {
			throw new Error('You must enter your identification number');
		};

		// Implement checking identity procedure (redirect the client to national database for example and verify identity)

		/**
		 * */ 

		// Verify if client fill areaCode box
		if (!asset.areaCode) {
			throw new Error('You must enter your areaCode, you can find it on https://skalleo.io');
		};

		// Implement checking areaCode procedure (redirect the client to national database for example and verify if this zone is concerned)
		// OR verify it with apply we can look CHAIN_STATE_POOL_ACCOUNTS if the areaCode is available

		/**
		 * */
		

		// Verify if username is valid
		const chunks = asset.username.split(/\./);

		if (chunks.length > 2) {
			throw new Error('You may not use "." than for the domain');
		};

		if (!VALID_PATIENT_DOMAIN.includes(chunks[1])) {
			throw new Error(`Invalid domain found "${chunks[1]}". Valid TLDs are "${VALID_PATIENT_DOMAIN.join()}"`);
		};

	

		

	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async apply({ asset, transaction, stateStore }: ApplyAssetContext<{}>): Promise<void> {
		//create Patient account
		const senderAddress = transaction.senderAddress;
		const senderAccount = await stateStore.account.get(senderAddress);
		const areaCode = asset.areaCode;

		//Admit only one patient account per account
		if (senderAccount.patient.selfPatient) {
			throw new Error('You have already a patient account !')
		}

		//Verify if areaCode is valid
		if (!(areaCode in CHAIN_STATE_POOL_ACCOUNTS)) {
			throw new Error('Your area is not yet concerned, sorry !');
		}

		const patientAccount = createPatientAccount({
			patientIdentificationNumber: asset.patientIdentificationNumber,
        	areaCode: asset.areaCode,
        	username: asset.username,
        	ownerAddress: asset.ownerAddress,
			nonce: asset.nonce,
		});

		//update sender account with unique Patient ID
		senderAccount.patient.selfPatient = patientAccount;
		await stateStore.account.set(senderAddress, senderAccount);
		
		//save patient
		const allAccounts = await getAllPatientAccounts(stateStore);
		allAccounts.push(patientAccount);
		await setAllPatientAccounts(stateStore, allAccounts);






	}



}





