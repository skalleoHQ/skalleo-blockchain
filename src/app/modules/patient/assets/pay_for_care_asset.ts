import { BaseAsset, ApplyAssetContext, ValidateAssetContext } from 'lisk-sdk';

export class PayForCareAsset extends BaseAsset {
	public name = 'payForCare';
  public id = 2;

  // Define schema for asset
	public schema = {
    $id: 'patient/payForCare-asset',
		title: 'PayForCareAsset transaction asset for patient module',
		type: 'object',
		required: [],
		properties: {},
  };

  public validate({ asset }: ValidateAssetContext<{}>): void {
    // Validate your asset
  }

	// eslint-disable-next-line @typescript-eslint/require-await
  public async apply({ asset, transaction, stateStore }: ApplyAssetContext<{}>): Promise<void> {
		throw new Error('Asset "payForCare" apply hook is not implemented.');
	}
}
