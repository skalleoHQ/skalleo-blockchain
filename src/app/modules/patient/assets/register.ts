export interface PatientModuleProps {
    selfPatient: 'bytes';
    reverseLookup: 'bytes';
};


export interface CreatePatientAccountAssetProps {
	patientIdentificationNumber: string;
	areaCode: string;
	username: string;
};


export interface RegisteredPatientAccountsSchemaProps {
	registeredPatientAccounts: []
};



