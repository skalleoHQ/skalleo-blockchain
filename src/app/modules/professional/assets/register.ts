export interface ProfessionalModuleProps{
    selfProfessional: Buffer,
    reverseLookup: Buffer
};


export interface CreateProfessionalAccountAssetProps {
    readonly professionalIdentificationNumber: string;
    readonly areaCode: string;
    readonly username: string;
};


export interface RegisteredProfessionalAccountsSchemaProps {
    readonly registeredProfessionalAccounts: [];
};


export interface RegisteredProfessionalAccount {
    readonly id: Buffer;
    readonly ownerAddress: Buffer;
    readonly professionalIdentificationNumber: string;
    readonly areaCode: string;
    readonly username: string;
}




