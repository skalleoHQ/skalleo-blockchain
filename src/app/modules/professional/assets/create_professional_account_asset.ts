import { BaseAsset, ApplyAssetContext, ValidateAssetContext } from 'lisk-sdk';

import { CreateProfessionalAccountAssetProps, ProfessionalModuleProps } from './register';

const { 
	VALID_PROFESSIONAL_DOMAIN,
	getAllProfessionalAccounts,
	createProfessionalAccount,
	setAllPatientAccounts,

} = require ('../data/utils')


export class CreateProfessionalAccountAsset extends BaseAsset<CreateProfessionalAccountAssetProps> {
	public name = 'createProfessionalAccount';
  	public id = 1;

  // Define schema for asset
	public schema = {
    $id: 'professional/createProfessionalAccount-asset',
		title: 'CreateProfessionalAccountAsset transaction asset for professional module',
		type: 'object',
		required: ['professionalIdentificationNumber', 'areaCode', 'username'],
		properties: {
			professionalIdentificationNumber: {
				dataType: 'string',
				fieldNumber: 1,
			},
			username: {
				dataType: 'string',
				fieldNumber: 2,
			},
			areaCode: {
				dataType: 'string',
				fieldNumber: 3,
			}
		},
  	};



	public validate({ asset }: ValidateAssetContext<CreateProfessionalAccountAssetProps>): void {
		// Verify if client fill professional Identification Number box
		if (!asset.professionalIdentificationNumber) {
			throw new Error('You must enter your identification number');
		};

		// Must implement checking identity procedure (redirect the client to his national database for example and verify identity)

		/**
		 * */ 

		// Verify if client fill areaCode box
		if (!asset.areaCode) {
			throw new Error('You must enter your areaCode, you can find it on https://skalleo.io');
		};

		// Implement checking areaCode procedure (redirect the client to his national database for example and verify if this zone is concerned)
		// OR verify it with apply: we can look CHAIN_STATE_POOL_ACCOUNTS if the areaCode is available

		/**
		 * */
		
		// Verify if professional fill username box
		if (!asset.username) {
			throw new Error('You must enter your username, please use "." only in front of "pro');
		};

		// Verify if username is valid
		const chunks = asset.username.split(/\./);

		if (chunks.length > 2) {
			throw new Error('You may not use "." than for the domain');
		};

		if (!VALID_PROFESSIONAL_DOMAIN.includes(chunks[1])) {
			throw new Error(`Invalid domain found "${chunks[1]}". Valid domain is "${VALID_PROFESSIONAL_DOMAIN.join()}"`);
		};
	}

	// eslint-disable-next-line @typescript-eslint/require-await
  	public async apply({ asset, transaction, stateStore }: ApplyAssetContext<CreateProfessionalAccountAssetProps>): Promise<void> {
		
		const senderAddress = transaction.senderAddress;
		const senderAccount = await stateStore.account.get<ProfessionalModuleProps>(senderAddress);

		const professionalAccounts = await getAllProfessionalAccounts(stateStore);

		//Admit only one professional account by the way
		if (senderAccount.selfProfessional || senderAccount.reverseLookup) {
			throw new Error('You have already a professional account !')
		}

		//Each professionalIdentificationNumber has only one account
		const professionalIdentificationNumberIndex = professionalAccounts.findIndex((t) => t.professionalIdentificationNumber === (asset.professionalIdentificationNumber));
		if (professionalIdentificationNumberIndex >= 0) {
			throw new Error('This professional identification number is already registered !')
		}

		//Verify if username is unique
		const professionalUsernameIndex = professionalAccounts.findIndex((t) => t.username === (asset.username));
		if (professionalUsernameIndex >= 0) {
			throw new Error('This username is already reserved, please try another.')
		}

		//Verify if areaCode is valid

		/*const areaCodeIndex = poolAccounts.findIndex((t) => t.areaCode === (asset.areaCode));
		if (areaCodeIndex >= 0) {
			throw new Error('Your area is not yet concerned, sorry !');
		}*/

		//create professional account
		const professionalAccount = createProfessionalAccount({
			professionalIdentificationNumber: asset.professionalIdentificationNumber,
        	areaCode: asset.areaCode,
        	username: asset.username,
        	ownerAddress: senderAddress,
			nonce: transaction.nonce,
		});

		//update user account with unique username and his personal professional account
		senderAccount.selfProfessional = professionalAccount;
		senderAccount.reverseLookup = professionalAccount.username;
		await stateStore.account.set(senderAddress, senderAccount);
		
		//save professional in database
		const allAccounts = await getAllProfessionalAccounts(stateStore);
		allAccounts.push(professionalAccount);
		await setAllPatientAccounts(stateStore, allAccounts);

	}

}




module.exports = {
	CreateProfessionalAccountAsset,
}
