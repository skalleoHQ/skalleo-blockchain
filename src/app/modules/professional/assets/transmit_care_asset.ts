import { BaseAsset, ApplyAssetContext, ValidateAssetContext } from 'lisk-sdk';
import { TransmitCareAssetProps } from './register';


const {
	VALID_PATIENT_DOMAIN,
	getAllPatientAccounts,

} = require ('../../patient/data/utils');



const {
	getAllProfessionalAccounts,

} = require ('./register');



const {
	recordCare,
	getAllRecordedCare,
	setAllRecordedCare,
	
} = require ('../data/care')



export class TransmitCareAsset extends BaseAsset<TransmitCareAssetProps> {
	public name = 'transmitCare';
  	public id = 2;

  // Define schema for asset
	public schema = {
    $id: 'professional/transmitCare-asset',
		title: 'TransmitCareAsset transaction asset for professional module',
		type: 'object',
		required: ['patientIdentificationNumber', 'reverseLookup', 'areaCode', 'careSpecifications'],
		properties: {
			patientIdentificationNumber: {
				dataType: 'string',
				fieldNumber: 1,
			},
			reverseLookup: {
				dataType: 'string',
				fieldNumber: 2,
			},
			areaCode: {
				dataType: 'string',
				fieldNumber: 3
			},
			careSpecifications: {
				dataType: 'string',
				fieldNumber: 4,
			},
		},
	};

	public validate({ asset }: ValidateAssetContext<TransmitCareAssetProps>): void {

		if (!asset.patientIdentificationNumber) {
			throw new Error('You must enter identification number of your patient');
		}

		if (!asset.reverseLookup) {
			throw new Error('You must enter username of your patient');
		}

		const chunks = asset.reverseLookup.split(/\./);

		if (chunks.length > 2) {
			throw new Error('You may not use "." than for the domain');
		};

		if (!VALID_PATIENT_DOMAIN.includes(chunks[1])) {
			throw new Error(`Invalid domain found "${chunks[1]}". Valid domain is "${VALID_PATIENT_DOMAIN.join()}"`);
		};

		if (!asset.areaCode) {
			throw new Error('You must enter areaCode of your patient');
		}

		//This careSpecifications will be used to update patient's medical record in his national database and plugin database
		if (!asset.careSpecifications) {
			throw new Error('Please update the cares specifications of your patient !')
		}

		//Must implement procedure in order to send specifications to national database & verify if careSpecs respect skeleton

	}

	// eslint-disable-next-line @typescript-eslint/require-await
  	public async apply({ asset, transaction, stateStore }: ApplyAssetContext<TransmitCareAssetProps>): Promise<void> {
		const senderAddress = transaction.senderAddress;
		
		//Verify if patient exists
		const patientAccounts = await getAllPatientAccounts(stateStore);

		const patientIdentificationNumberIndex = patientAccounts.findIndex((t) => t.patientIdentificationNumber === asset.patientIdentificationNumber);
		if (patientIdentificationNumberIndex < 0) {
			throw new Error('Patient not found !');	
		}

		const patientReverseLookupIndex = patientAccounts.findIndex((t) => t.username === asset.reverseLookup);
		if (patientReverseLookupIndex < 0) {
			throw new Error('Patient not found !');
		}

		/*const patientAreaCodeIndex = patientAccounts.findIndex((t) => t.areaCode === asset.areaCode);
		if (patientAreaCodeIndex < 0) {
			throw new Error('Patient not found !');
		}*/
		//assert(patientReverseLookupIndex === patientAreaCodeIndex); impossible because areaCode not unique

		if(patientIdentificationNumberIndex !== patientReverseLookupIndex) {
			throw new Error('Patient not found !');
		}

		const professionalAccounts = await getAllProfessionalAccounts(stateStore);
		const addressIndex = professionalAccounts.findIndex((t) => t.ownerAddress === transaction.senderAddress);
		
		const professionalAccount = professionalAccounts[addressIndex];
		const patientAccount = patientAccounts[patientIdentificationNumberIndex]; 

		if (professionalAccount.ownerAddress === patientAccount.ownerAddress) {
			throw new Error('Transaction error !');
		}
		
		//create care for a patient
		const care = recordCare({
			patientIdentificationNumberIndex: asset.patientIdentificationNumber,
			reverseLookup: asset.reverseLookup,
			areaCode: asset.areaCode,
			careSpecifications: asset.careSpecifications,
			senderAddress: senderAddress,
			nonce: transaction.nonce,
		})

		//update care data chain
		const allCare = await getAllRecordedCare(stateStore);
		allCare.push(care);
		await setAllRecordedCare(stateStore, allCare);

	}


}



module.exports = {
	TransmitCareAsset,
}
