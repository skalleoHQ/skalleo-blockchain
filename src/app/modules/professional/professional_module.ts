/* eslint-disable class-methods-use-this */

import {
    AfterBlockApplyContext,


    AfterGenesisBlockApplyContext, BaseModule,


    BeforeBlockApplyContext, TransactionApplyContext
} from 'lisk-sdk';

import { CreateProfessionalAccountAsset } from "./assets/create_professional_account_asset";

const { EMPTY_BUFFER, 

} = require ('../professional/data/utils');



export class ProfessionalModule extends BaseModule {
    public accountSchema = {
        $id: 'skalleo/professional/professionalAccount',
        type: 'object',
        required: ['selfProfessional', 'reverseLookup'],
        properties: {
            reverseLookup: {
                dataType: 'bytes',
                fieldNumber: 1,
            },
            selfProfessional: {
                dataType: 'bytes',
                fieldNumber: 2,
            },
        },
        default: {
            selfProfessional: EMPTY_BUFFER,
            reverseLookup: EMPTY_BUFFER,
        }
    };


    public actions = {
        // Example below
        // getBalance: async (params) => this._dataAccess.account.get(params.address).token.balance,
        // getBlockByID: async (params) => this._dataAccess.blocks.get(params.id),
    };
    public reducers = {
        // Example below
        // getBalance: async (
		// 	params: Record<string, unknown>,
		// 	stateStore: StateStore,
		// ): Promise<bigint> => {
		// 	const { address } = params;
		// 	if (!Buffer.isBuffer(address)) {
		// 		throw new Error('Address must be a buffer');
		// 	}
		// 	const account = await stateStore.account.getOrDefault<TokenAccount>(address);
		// 	return account.token.balance;
		// },
    };
    public name = 'professional';
    public transactionAssets = [new CreateProfessionalAccountAsset()];
    public events = [
        // Example below
        // 'professional:newBlock',
    ];
    public id = 1001;

    // public constructor(genesisConfig: GenesisConfig) {
    //     super(genesisConfig);
    // }

    // Lifecycle hooks
    public async beforeBlockApply(_input: BeforeBlockApplyContext) {
        // Get any data from stateStore using block info, below is an example getting a generator
        // const generatorAddress = getAddressFromPublicKey(_input.block.header.generatorPublicKey);
		// const generator = await _input.stateStore.account.get<TokenAccount>(generatorAddress);
    }

    public async afterBlockApply(_input: AfterBlockApplyContext) {
        // Get any data from stateStore using block info, below is an example getting a generator
        // const generatorAddress = getAddressFromPublicKey(_input.block.header.generatorPublicKey);
		// const generator = await _input.stateStore.account.get<TokenAccount>(generatorAddress);
    }

    public async beforeTransactionApply(_input: TransactionApplyContext) {
        // Get any data from stateStore using transaction info, below is an example
        // const sender = await _input.stateStore.account.getOrDefault<TokenAccount>(_input.transaction.senderAddress);
    }

    public async afterTransactionApply(_input: TransactionApplyContext) {
        // Get any data from stateStore using transaction info, below is an example
        // const sender = await _input.stateStore.account.getOrDefault<TokenAccount>(_input.transaction.senderAddress);
    }

    public async afterGenesisBlockApply(_input: AfterGenesisBlockApplyContext) {
        // Get any data from genesis block, for example get all genesis accounts
        // const genesisAccounts = genesisBlock.header.asset.accounts;
    }
}
