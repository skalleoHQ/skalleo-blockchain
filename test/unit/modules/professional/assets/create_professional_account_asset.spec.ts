import { StateStore, testing } from 'lisk-sdk';
import { CreateProfessionalAccountAsset } from '../../../../../src/app/modules/professional/assets/create_professional_account_asset';
import { ProfessionalModuleProps } from '../../../../../src/app/modules/professional/assets/register';
import { ProfessionalModule } from '../../../../../src/app/modules/professional/professional_module';

describe('CreateProfessionalAccountAsset', () => {
  let transactionAsset: CreateProfessionalAccountAsset;

	beforeEach(() => {
		transactionAsset = new CreateProfessionalAccountAsset();
	});

	describe('constructor', () => {
		it('should have valid id', () => {
			expect(transactionAsset.id).toEqual(1);
		});

		it('should have valid name', () => {
			expect(transactionAsset.name).toEqual('createProfessionalAccount');
		});

		it('should have valid schema', () => {
			expect(transactionAsset.schema).toMatchSnapshot();
		});
	});

	describe('validate', () => {
		describe('schema validation', () => {
			it('should if username has more than one domain', () => {
				const context = testing.createValidateAssetContext({
					asset: { professionalIdentificationNumber: "3301012022", areaCode: "33_FRANCE", username: "PharmaciePorteD.Aix.pro" },
					transaction: { senderAddress: Buffer.alloc(0) } as any,
				})

				expect(() => transactionAsset.validate(context)).toThrow(
					'You may not use "." than for the domain'
				)
			});

			it('should throw error if username doesnt contain pro', () => {
				const context = testing.createValidateAssetContext({
					asset: { professionalIdentificationNumber: "3301012022", areaCode: "33_FRANCE", username: "PharmaciePorteD.Aix" },
					transaction: { senderAddress: Buffer.alloc(0) } as any,
				})

				expect(() => transactionAsset.validate(context)).toThrow(
					'Invalid domain found "Aix". Valid domain is "pro"'
				)
			});

			it('should throw error if user doesnt add his professional identification number', () => {
				const context = testing.createValidateAssetContext({
					asset: { professionalIdentificationNumber: "", areaCode: "33_FRANCE", username: "PharmaciePorteDAix.pro" },
					transaction: { senderAddress: Buffer.alloc(0) } as any,
				})

				expect(() => transactionAsset.validate(context)).toThrow(
					'You must enter your identification number'
				)
			});

			it('should throw error if user doesnt add his area code', () => {
				const context = testing.createValidateAssetContext({
					asset: { professionalIdentificationNumber: "3301012022", areaCode: "", username: "PharmaciePorteDAix.pro" },
					transaction: { senderAddress: Buffer.alloc(0) } as any,
				})

				expect(() => transactionAsset.validate(context)).toThrow(
					'You must enter your areaCode, you can find it on https://skalleo.io'
				)
			});

			it('should throw error if username doesnt a username', () => {
				const context = testing.createValidateAssetContext({
					asset: { professionalIdentificationNumber: "3301012022", areaCode: "33_FRANCE", username: "" },
					transaction: { senderAddress: Buffer.alloc(0) } as any,
				})

				expect(() => transactionAsset.validate(context)).toThrow(
					'You must enter your username, please use "." only in front of "pro'
				)
			});

			it('should be ok for valid schema', () => {
				const context = testing.createValidateAssetContext({
					asset: { professionalIdentificationNumber: "3301012022", areaCode: "33_FRANCE", username: "PharmaciePorteDAix.pro" },
					transaction: { senderAddress: Buffer.alloc(0) } as any,
				})

				expect(() => transactionAsset.validate(context)).not.toThrow();
			});

		});
	});

	describe('apply', () => {
		let stateStore: StateStore;
		let account: any;
		let account1: any;
		let account2: any;
		let account3: any;
		let account4: any;
		let account5: any


		beforeEach(() => {
			account = testing.fixtures.createDefaultAccount<ProfessionalModuleProps>([ProfessionalModule]);
			account1 = testing.fixtures.createDefaultAccount<ProfessionalModuleProps>([ProfessionalModule]);
			account2 = testing.fixtures.createDefaultAccount<ProfessionalModuleProps>([ProfessionalModule]);
			account3 = testing.fixtures.createDefaultAccount<ProfessionalModuleProps>([ProfessionalModule]);
			account4 = testing.fixtures.createDefaultAccount<ProfessionalModuleProps>([ProfessionalModule]);
			account5 = testing.fixtures.createDefaultAccount<ProfessionalModuleProps>([ProfessionalModule]);

			stateStore = new testing.mocks.StateStoreMock({
				accounts: [account, account1, account2, account3, account4, account5],
			});

			jest.spyOn(stateStore.chain, 'get');
			jest.spyOn(stateStore.chain, 'set');
		});

		describe('valid cases', () => {
			it('should update the state store with professional account', async () => {
				const context = testing.createApplyAssetContext({
					stateStore,
					asset: { professionalIdentificationNumber: "3301012022", areaCode: "33_FRANCE", username: "PharmaciePorteDAix.pro" },
					transaction: { senderAddress: account.address } as any,
				});

				await transactionAsset.apply(context);

				expect(stateStore.chain.networkIdentifier).toEqual(context.stateStore.chain.networkIdentifier);
				expect(stateStore.chain.set).toBe(context.stateStore.chain.set);
				expect(stateStore.chain.set).toHaveBeenCalledWith("professional: registeredProfessionalAccounts", expect.any(Buffer))
			});
		});

		describe('invalid cases', () => {
			it('should throw error if professional is already registered', async() => {
				const context = testing.createApplyAssetContext({
					stateStore,
					asset: { professionalIdentificationNumber: "3301012022", areaCode: "221_SENEGAL", username: "CliniqueNdiaye.pro" },
					transaction: { senderAddress: account.address } as any,
				});

				const context1 = testing.createApplyAssetContext({
					stateStore,
					asset: { professionalIdentificationNumber: "03301012022", areaCode: "221_SENEGAL", username: "OptiqueSud.pro" },
					transaction: { senderAddress: account.address } as any,
				});

				await transactionAsset.apply(context);

				await expect(() => transactionAsset.apply(context1)).rejects.toThrow(
					'You have already a professional account !'
				)

			});

			it('should throw error if professional identification number is already registered', async() => {
				const context2 = testing.createApplyAssetContext({
					stateStore,
					asset: { professionalIdentificationNumber: "003301012022", areaCode: "221_SENEGAL", username: "PharmacieBourguyba.pro" },
					transaction: { senderAddress: account2.address } as any,
				});

				const context3 = testing.createApplyAssetContext({
					stateStore,
					asset: { professionalIdentificationNumber: "003301012022", areaCode: "221_SENEGAL", username: "PharmacieAlmadies.pro" },
					transaction: { senderAddress: account3.address } as any,
				});

				await transactionAsset.apply(context2);

				await expect(() => transactionAsset.apply(context3)).rejects.toThrow(
					'This professional identification number is already registered !'
				)
			});

			it('should throw error if username is already reserved', async() => {
				const context4 = testing.createApplyAssetContext({
					stateStore,
					asset: { professionalIdentificationNumber: "0003301012022", areaCode: "221_SENEGAL", username: "CliniqueNdiaye.pro" },
					transaction: { senderAddress: account4.address } as any,
				});

				const context5 = testing.createApplyAssetContext({
					stateStore,
					asset: { professionalIdentificationNumber: "00003301012022", areaCode: "33_France", username: "CliniqueNdiaye.pro" },
					transaction: { senderAddress: account5.address } as any,
				});

				await transactionAsset.apply(context4);

				await expect(() => transactionAsset.apply(context5)).rejects.toThrow(
					'This username is already reserved, please try another.'
				)

			});

		});
	});
});
