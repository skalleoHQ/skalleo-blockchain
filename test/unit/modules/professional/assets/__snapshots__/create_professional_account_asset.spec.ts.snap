// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateProfessionalAccountAsset constructor should have valid schema 1`] = `
Object {
  "$id": "professional/createProfessionalAccount-asset",
  "properties": Object {
    "areaCode": Object {
      "dataType": "string",
      "fieldNumber": 3,
    },
    "professionalIdentificationNumber": Object {
      "dataType": "string",
      "fieldNumber": 1,
    },
    "username": Object {
      "dataType": "string",
      "fieldNumber": 2,
    },
  },
  "required": Array [
    "professionalIdentificationNumber",
    "areaCode",
    "username",
  ],
  "title": "CreateProfessionalAccountAsset transaction asset for professional module",
  "type": "object",
}
`;
